src/contrib/mmap_cache.rst

```diff
@@ -0,0 +1,506 @@
+    };
+
+The ``major_vno`` and ``minor_vno`` are used to tell the client about how the
+cache and the data is organized. With ``status`` SSSD can tell the client if the
+cache is valid or if e.g. the cache is recycled. In the latter case the client
+must close the current cache file and open a newly created.
+
+``b1`` and ``b2`` are used to implement a full memory barrier with the help of
+GCC's ``__sync_synchronize()`` built-in function. The memory barrier is used to
+make sure the header data can be updated without other processes interfering.
+
+``seed`` is a random seed for the hash function used to create lookup hashes
+from the provided name or ID. It is generated randomly whenever the cache file
+is created.
+
+``dt_size``, ``ft_size`` and ``ht_size`` are the sizes of the remaining three
+sections of the memory mapped cache and ``data_table``, ``free_table`` and
+``hash_table`` are the starting points of the sections relative to the start of
+the header.
+
+The Data Table
+--------------
+
+This is the main area to store the user and group data. It is organized in slots
+with a defined size. The current slot size is 40 bytes. Since the slot size is
+not given in the header the ``major_vno`` or ``minor_vno`` must be changed if
+the slot size is changed. Each data record for an individual user or group (see
+below) can occupy one or more consecutive slots.
+
+At startup the data table is initialized with ``0xff`` (all bits set) which is
+treated as invalid value by the memory cache code independent if 1, 4 or 8 bytes
+are read.
+
+The Free Table
+--------------
+
+The memory area is treated as bit-field. Each bit corresponds to a slot in the
+data table starting with the highest bit (``0x80``) of the first byte in the
+free table.
+
+At startup the free table is initialized with ``0x00`` (no bit set) which
+corresponds to a completely empty data table.
+
+The Hash Table
+--------------
+
+The hash table it treated as an array of ``uint32_t`` values. The index used to
+access the array is the hash value calculated from the user or group name or ID
+and the hash value is the slot number in the data table where the data record
+for the object starts.
+
+At startup the hash table is initialized with ``0xff`` which indicates that
+there is no data available for the given hash.
+
+Sizes of the tables
+-------------------
+
+From the previous sections it became clear that although the sizes of the data,
+free and hash tables are handled independently they are connected. Of course the
+free table should have at least so many bit as there are slots in the data
+table. But having more would be a waste of space (although not as much as having
+lesser bits). So number of bits and slots should match.
+
+The size of the data table should be large enough to help the amount of expected
+data where expected do not necessarily mean e.g. all possible users but the
+amount of data needed on the host running SSSD to work efficiently. It should be
```

Comment:
This first sentence is poorly worded, I guess it should be something like:

"The size of the data should be large enough to contain the expected data(amount of data needed on the host running SSSD to work efficiently).